<!DOCTYPE html><html lang="zh-CN"><head><meta name="generator" content="Hexo 3.9.0"><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"><meta name="description" content><meta name="keywords" content><meta name="author" content="Rr-shan,undefined"><meta name="copyright" content="Rr-shan"><title>【Sanzzi】</title><link rel="stylesheet" href="/css/fan.css"><link rel="stylesheet" href="/css/thirdparty/jquery.mCustomScrollbar.min.css"><link rel="icon" href="/favicon.ico"><!-- script(src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML")--><script src="/js/mathjax/mathjax.js"></script><script type="text/x-mathjax-config">MathJax.Hub.Config({
    tex2jax: {inlineMath: [['$', '$'], ['\\(', '\\)']]}
});
</script><script>var isPassword = '' || false;
if (isPassword) {
    if (prompt('请输入文章密码') !== '') {
        alert('密码错误！');
        history.back();
    }
}</script><script>window.GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  }
}</script></head><body><canvas id="universe"></canvas><!--#body--><div id="sidebar"><div class="author-info"><div class="author-info-avatar"><img class="author-info-avatar-img" src="/avatar.png"></div><div class="author-info-name">Rr-shan</div><div class="author-info-description"></div><div class="links-buttons"><a class="links-button button-hover" href="https://github.com/Rr-shan" target="_blank">GitHub<i class="icon-dot bg-color9"></i></a><a class="links-button button-hover" href="mailto:1224559633@qq.com" target="_blank">E-Mail<i class="icon-dot bg-color6"></i></a></div><div class="author-info-articles"><a class="author-info-articles-archives article-meta" href="archives"><span class="pull-top">日志</span><span class="pull-bottom">40</span></a><a class="author-info-articles-tags article-meta" href="tags"><span class="pull-top">标签</span><span class="pull-bottom">11</span></a><a class="author-info-articles-categories article-meta" href="categories"><span class="pull-top">分类</span><span class="pull-bottom">12</span></a></div><div class="friend-link"><a class="friend-link-text" href="https://chao-yin-github.github.io/" target="_blank">yccccc~~~~~~~~</a><a class="friend-link-text" href="https://sanzzi.cn" target="_blank">Mywebsite</a></div></div></div><div id="main-container"><header><div id="menu-outer"><i class="menu-list-icon fas fa-bars"></i><nav id="menu-inner"><a class="menu-item" href="/">首页</a><a class="menu-item" href="/tags">标签</a><a class="menu-item" href="/categories">分类</a><a class="menu-item" href="/archives">归档</a><a class="menu-item" href="/about">关于</a></nav><div class="right-info"><a class="title-name" href="/">Sanzzi</a><span id="now-time"></span></div></div></header><div id="content-outer"><div id="content-inner"><div id="recent-posts"><!-- each post in page.posts.sort('date', -1).limit(10).toArray()--><!-- config中配置按照什么排序--><div class="recent-post-item"><a class="post-title" href="2020/04/05/生涯规划课之如何写一份优秀的简历/">生涯规划课之如何写一份优秀的简历</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-05</time></div><div class="post-content"><div class="main-content content">已被加密啦</div></div><a class="button-hover more" href="2020/04/05/生涯规划课之如何写一份优秀的简历/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/04/05/解析二级域名后看是否生效/">解析二级域名后看是否生效</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-05</time></div><div class="post-content"><div class="main-content content"><h5 id="阿里云的二级域名解析："><a href="#阿里云的二级域名解析：" class="headerlink" title="阿里云的二级域名解析："></a>阿里云的二级域名解析：</h5><p><strong>解析线路：</strong>选择 默认 （默认为必选项，如未设置会导致部分用户无法访问 )。<br><strong>TTL：</strong>为缓存时间，数值越小，<strong>修改记录各地生效时间越快</strong>，默认为10分钟。</p>
<figure class="highlight cmd"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">ping</span> yc.sanzzi.cn</span><br></pre></td></tr></table></figure>

<p>如果生效是以下这样的</p>
<figure class="highlight cmd"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">正在 <span class="built_in">Ping</span> yc.sanzzi.cn [<span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span>] 具有 <span class="number">32</span> 字节的数据:</span><br><span class="line">来自 <span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span> 的回复: 字节=<span class="number">32</span> 时间=<span class="number">43</span>ms TTL=<span class="number">49</span></span><br><span class="line">来自 <span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span> 的回复: 字节=<span class="number">32</span> 时间=<span class="number">43</span>ms TTL=<span class="number">49</span></span><br><span class="line">来自 <span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span> 的回复: 字节=<span class="number">32</span> 时间=<span class="number">308</span>ms TTL=<span class="number">49</span></span><br><span class="line">来自 <span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span> 的回复: 字节=<span class="number">32</span> 时间=<span class="number">170</span>ms TTL=<span class="number">49</span></span><br><span class="line"></span><br><span class="line"><span class="number">123</span>.<span class="number">57</span>.<span class="number">242</span>.<span class="number">220</span> 的 <span class="built_in">Ping</span> 统计信息:</span><br><span class="line">    数据包: 已发送 = <span class="number">4</span>，已接收 = <span class="number">4</span>，丢失 = <span class="number">0</span> (<span class="number">0</span>% 丢失)，</span><br><span class="line">往返行程的估计时间(以毫秒为单位):</span><br><span class="line">    最短 = <span class="number">43</span>ms，最长 = <span class="number">308</span>ms，平均 = <span class="number">141</span>ms</span><br></pre></td></tr></table></figure>

<p>失败是</p>
<figure class="highlight cmd"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">Ping</span> 请求找不到主机 sanzi.sanzzi.cn。请检查该名称，然后重试。</span><br></pre></td></tr></table></figure>

</div></div><a class="button-hover more" href="2020/04/05/解析二级域名后看是否生效/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/04/03/JShell容易出现错误之处/">JShell容易出现错误之处</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-05</time><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="categories/Java/">Java</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="tags/Java/">Java</a></div></div><div class="post-content"><div class="main-content content"><h3 id="注意事项-amp-可能存在的问题"><a href="#注意事项-amp-可能存在的问题" class="headerlink" title="注意事项&amp;可能存在的问题"></a>注意事项&amp;可能存在的问题</h3><ul>
<li>不要调⽤<strong>Runtime类</strong>直接使⽤<strong>系统的shell命令</strong></li>
<li>编码要求使⽤<strong>uft-8</strong>编码</li>
<li>echo、grep没有处理 \n<strong>转义字符的问题</strong></li>
<li>可能程序中使⽤switch语句进⾏流程控制，但是只取输⼊前两位达到省事的⽬的，这样是不可取的，就好⽐，别⼈输⼊pw，程序应当提⽰找不到pw命令,而⾮直接调⽤pwd命令。</li>
<li>整个代码只有⼀个类，并在类⾥使⽤内部类，代码<strong>耦合程度⾼</strong>，而且难以阅读。功能存在问题，未经测试就匆忙提交，质量普遍不⾼。</li>
<li><strong>接口没有打注释</strong></li>
<li>请正确认识<strong>交互性</strong>，不要给程序运⾏添加不必要的操作。</li>
<li><strong>不要把⽂件路径写死了</strong></li>
<li>这是⼀个功能简化了的shell，但基本忠于shell，不要乱改原有的命令。</li>
<li><strong>命名随便</strong>，有⽤a,b,c的，还有竟然⽤⾃⼰名字的。</li>
<li>尽量不要使⽤<strong>windows编程</strong></li>
<li>请把的<strong>注释掉的代码删去</strong>，需要保留的说明原因。</li>
</ul>
<p>From:yc<br><em>感觉犯错点很有参考价值</em></p>
</div></div><a class="button-hover more" href="2020/04/03/JShell容易出现错误之处/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/04/03/论文阅读DrImpute/">论文阅读DrImpute</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-05</time><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="categories/ML-DL/">ML&amp;DL</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="tags/Dropouts课题-刘世超老师/">Dropouts课题(刘世超老师)</a></div></div><div class="post-content"><div class="main-content content"><h2 id="原理"><a href="#原理" class="headerlink" title="原理"></a>原理</h2><h4 id="聚类"><a href="#聚类" class="headerlink" title="聚类"></a>聚类</h4><h4 id="1-preprocess：data-cleansing-normalization-and-log-transformation"><a href="#1-preprocess：data-cleansing-normalization-and-log-transformation" class="headerlink" title="1. preprocess：data cleansing, normalization, and log transformation"></a>1. <strong>preprocess：</strong>data cleansing, normalization, and log transformation</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">preprocess &lt;- <span class="keyword">function</span>(x, min.expressed.gene = <span class="number">0</span>, min.expressed.cell = <span class="number">2</span>, max.expressed.ratio = <span class="number">1</span>, normalize.by.size.effect = <span class="literal">FALSE</span>)&#123;</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> (class(x) == <span class="string">'SummarizedExperiment'</span>)</span><br><span class="line">		X &lt;- assays(x)$count</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (class(x) == <span class="string">'matrix'</span>)</span><br><span class="line">		X &lt;- x</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (is(x, <span class="string">'sparseMatrix'</span>))</span><br><span class="line">		X &lt;- x</span><br><span class="line">	<span class="keyword">else</span></span><br><span class="line">		<span class="keyword">stop</span>(sprintf(<span class="string">'unknown class(x): %s'</span>, class(x)))</span><br><span class="line"></span><br><span class="line">	M &lt;- ncol(X)</span><br><span class="line">	N &lt;- nrow(X)</span><br><span class="line">	m &lt;- Matrix::colSums(X &gt; <span class="number">1</span>) &gt;= min.expressed.gene	<span class="comment"># cells that have at least min.expressed.gene expreseed genes</span></span><br><span class="line">	n &lt;- Matrix::rowSums(X &gt; <span class="number">1</span>) &lt;= max.expressed.ratio * M &amp; Matrix::rowSums(X &gt; <span class="number">1</span>) &gt;= min.expressed.cell	<span class="comment"># genes that are detected in at least min.expressed.cell or at most max.expressed.ratio cells</span></span><br><span class="line">	<span class="keyword">if</span> (normalize.by.size.effect)&#123;</span><br><span class="line">	  sf &lt;- apply((X[n, m] + <span class="number">1</span>) / exp(Matrix::rowMeans(log(X[n, m] + <span class="number">1</span>))), <span class="number">2</span>, median)</span><br><span class="line">		X &lt;- t(t(X[n, m]) / sf)</span><br><span class="line">	&#125;<span class="keyword">else</span></span><br><span class="line">		X &lt;- X[n, m]</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> (class(x) == <span class="string">'SummarizedExperiment'</span>)&#123;</span><br><span class="line">		x &lt;- x[n, m]</span><br><span class="line">		assays(x)$count &lt;- X</span><br><span class="line">	&#125;<span class="keyword">else</span> <span class="keyword">if</span> (class(x) == <span class="string">'matrix'</span>)&#123;</span><br><span class="line">		x &lt;- as.matrix(X)</span><br><span class="line">	&#125;<span class="keyword">else</span> <span class="keyword">if</span> (is(x, <span class="string">'sparseMatrix'</span>))&#123;</span><br><span class="line">		x &lt;- X</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	x</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="2-calculating-the-distance-matrix-among-cells"><a href="#2-calculating-the-distance-matrix-among-cells" class="headerlink" title="2.  calculating the distance matrix among cells"></a>2.  calculating the distance matrix among cells</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">dist2 &lt;- <span class="keyword">function</span>(X, method)&#123;</span><br><span class="line">	<span class="keyword">if</span> (method == <span class="string">'spearman'</span>)</span><br><span class="line">		D &lt;- as.matrix(<span class="number">1</span> - cor(as.matrix(X), method = <span class="string">'spearman'</span>))</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (method == <span class="string">"pearson"</span>)</span><br><span class="line">		D &lt;- as.matrix(<span class="number">1</span> - cor(as.matrix(X), method = <span class="string">'pearson'</span>))</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (method == <span class="string">"euclidean"</span>) </span><br><span class="line">		D &lt;- as.matrix(dist(t(as.matrix(X)), method = <span class="string">'euclidean'</span>))</span><br><span class="line">	<span class="keyword">else</span></span><br><span class="line">		<span class="keyword">stop</span>(sprintf(<span class="string">'unknown method: %s'</span>, method))</span><br><span class="line">	D</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="3-imputing-the-dropout-entries-based-on-the-clustering-results"><a href="#3-imputing-the-dropout-entries-based-on-the-clustering-results" class="headerlink" title="3.  imputing the dropout entries based on the clustering results"></a>3.  imputing the dropout entries based on the clustering results</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">DrImpute &lt;- <span class="keyword">function</span>(X, ks = <span class="number">10</span>:<span class="number">15</span>, dists = c(<span class="string">'spearman'</span>, <span class="string">'pearson'</span>), batch.size = <span class="literal">NA</span>, mc.cores = <span class="number">1</span>)&#123;</span><br><span class="line"></span><br><span class="line">	N &lt;- nrow(X)	<span class="comment"># number of genes</span></span><br><span class="line">	M &lt;- ncol(X) 	<span class="comment"># number of cells</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] number of cells: %d\n'</span>, Sys.time(), M))</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] number of genes: %d\n'</span>, Sys.time(), N))</span><br><span class="line"></span><br><span class="line">	is.zero &lt;- as(X == <span class="number">0</span>, <span class="string">'lgeMatrix'</span>)	<span class="comment"># dense logical matrix for zero and weakly detected events # 零和弱检测事件的密集逻辑矩阵</span></span><br><span class="line"></span><br><span class="line">	is.expressed &lt;- as(X &gt; log(<span class="number">1</span> + <span class="number">5</span>), <span class="string">'lgeMatrix'</span>) <span class="comment"># dense logical matrix for robustly expressed events # 鲁棒表示事件的密集逻辑矩阵</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% zero events (X=0): %.1f%%\n'</span>, Sys.time(), sum(is.zero) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% robustly expressed events (X&gt;5): %.1f%%\n'</span>, Sys.time(), sum(is.expressed) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	<span class="keyword">if</span> (!is.na(batch.size))&#123;</span><br><span class="line">		cat(sprintf(<span class="string">'[%s] batch size: %d cells\n'</span>, Sys.time(), batch.size))</span><br><span class="line">	&#125;</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] # of CPU cores: %d\n'</span>, Sys.time(), mc.cores))</span><br><span class="line"></span><br><span class="line">	X &lt;- estimate.expression(X, ks = ks, dists = dists, mc.cores = mc.cores, batch.size = batch.size)</span><br><span class="line">	is.zero &lt;- as(X == <span class="number">0</span>, <span class="string">'lgeMatrix'</span>)	<span class="comment"># dense logical matrix for zero indicators</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% zero events after imputation: %.1f%%\n'</span>, Sys.time(), sum(is.zero) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	X</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="4-averaging-all-imputation-results-to-determine-the-final-imputation"><a href="#4-averaging-all-imputation-results-to-determine-the-final-imputation" class="headerlink" title="4.  averaging all imputation results to determine the final imputation."></a>4.  averaging all imputation results to determine the final imputation.</h4><h2 id="实验"><a href="#实验" class="headerlink" title="实验"></a>实验</h2><h3 id="Experiment-1："><a href="#Experiment-1：" class="headerlink" title="Experiment 1："></a>Experiment 1：</h3><p><strong>Method:</strong><br>the <strong>cell-cell</strong> distance matrix was computed using <strong>Spearman and Pearson correlations</strong>, followed by the <strong>cell-wise clustering</strong> based upon the distance matrix over a range of expected number of clusters k (k ranging from 10 to 15 by default).</p>
<p><strong>Evaluation indicators:</strong><br>To generate the dropout zero, we randomly <strong>down-sampled</strong> the raw sequencing reads to 10, 15, 25, 40 and 63%f the total number of reads</p>
<p><strong>Evaluation indicators:<br>F1 score</strong></p>
<h3 id="Experiment-2"><a href="#Experiment-2" class="headerlink" title="Experiment 2"></a>Experiment 2</h3><p><strong>Method:</strong><br>The effect of cell clustering.<br> For each dataset, we randomly sampled 100 genes (<strong>down-sampling</strong>), or one-third of the total cells (down-sampling), and we clustered the cells using each of the clustering methods with and without preprocessing the down-sampled dataset using DrImpute. This process was <strong>repeated 100</strong> times, and we compared how consistent the clustering results were after down-sampling the genes or cells as measured by cross ARI<br><strong>Evaluation indicators:<br>ARI</strong></p>
<h3 id="Experiment-3"><a href="#Experiment-3" class="headerlink" title="Experiment 3"></a>Experiment 3</h3><p><strong>Method:</strong><br>To evaluate the accuracy of the <strong>dimension reduction</strong> in 2D space, we first estimated how discriminatively the cells from one population (using the class label reported in the original publication) separated from other populations in 2D space. For each dimension reduction result</p>
<p><strong>Evaluation indicators:</strong><br>we used the 2D coordinates of 90% of cells as the feature to train a linear support vector machine (SVM) classifier,and we predicted the class label for the remaining 10%of the cells. The above process was repeated ten times,and the overall prediction accuracy (10-fold cross validation accuracy) was used to quantitatively measure the separation of different populations in 2D space.// 这块没有讲清楚</p>
<h3 id="Experiment-4"><a href="#Experiment-4" class="headerlink" title="Experiment 4"></a>Experiment 4</h3><p>重建谱系轨迹，推断单细胞的分化和祖细胞状态。(老师也不是很明白)</p>
<h2 id="数据集："><a href="#数据集：" class="headerlink" title="数据集："></a>数据集：</h2><p>scRNA-seq datasets</p>
<p><strong>cell labels</strong><br>“gold”：experimental conditions or cell lines（细胞系）<br>“silver”： derived and assigned by the authors’ knowledge of the underlying biology.<br>“bronze”:  developmental stages (time labels)</p>
</div></div><a class="button-hover more" href="2020/04/03/论文阅读DrImpute/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/04/03/DeepImpute/">论文阅读DeepImpute</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-05</time><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="categories/ML-DL/">ML&amp;DL</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="tags/Dropouts课题-刘世超老师/">Dropouts课题(刘世超老师)</a></div></div><div class="post-content"><div class="main-content content"><h2 id="研究背景"><a href="#研究背景" class="headerlink" title="研究背景"></a>研究背景</h2><ol>
<li>RNA测序技术发展，通过<strong>转录组水平研究基因</strong>；</li>
<li>10X genomics公司，测序scRNA（ Fluidigm- and Drop-Seq-based methods ）；</li>
<li><strong>问题：</strong>每个细胞的<strong>mRNA拷贝起始量较低，因此scRNA序列数据非常嘈杂且不完整</strong>；</li>
<li>在一个scRNA序列实验中，经常观察到缺失值<strong>超过70%</strong>的数据集；</li>
<li><strong>零值的可能性：</strong>a.真零值；b.<strong>dropout</strong>，原始RNA转录物未能扩增所致；</li>
</ol>
<h2 id="已有的插补算法"><a href="#已有的插补算法" class="headerlink" title="已有的插补算法"></a>已有的插补算法</h2><ul>
<li><strong>MAGIC</strong>：建立<strong>马尔可夫转移矩阵</strong>并<strong>平滑数据</strong></li>
<li><strong>ScImpute</strong>：建立<strong>LASSO回归模型</strong>，对每个细胞迭代地进行补全/估算</li>
<li><strong>SAVER</strong>：是一个<strong>使用各种先验概率函数</strong>的<strong>基于贝叶斯的模型</strong></li>
<li><strong>DrImpute</strong>：是一种<strong>基于聚类</strong>的方法，它使用一致的策略:它用<strong>几个类的先验矩阵或距离矩阵</strong>估计一个值，然后用聚集法进行补全/估算</li>
<li><strong>VIPER</strong>：它着眼于<strong>细胞/细胞</strong>之间的相互作用，以适应每个细胞的<strong>线性模型</strong></li>
<li><strong>DCA</strong>：DCA构建了一个<strong>自动编码器</strong>，使用一个<strong>零膨胀负二项先验</strong>对<strong>基因分布</strong>进行建模，自动编码器试图预测基因的<strong>均值</strong>、<strong>标准差</strong>和<strong>丢失概率</strong>。但这些方法大多数都需要大量的运行时间，对更快，和可扩展的补全方法的需求也在不断增加。</li>
<li><strong>DeepImpute</strong>：<strong>分而治之</strong>的策略，构建多个亚/子神经网络，每个子神经网络都由四层组成。</li>
</ul>
<h2 id="DeepImpute的工作流程"><a href="#DeepImpute的工作流程" class="headerlink" title="DeepImpute的工作流程"></a>DeepImpute的工作流程</h2><ol>
<li>preprocessing</li>
<li>Architecture</li>
<li>training procedure</li>
<li>imputation<h3 id="Preprocessing预处理"><a href="#Preprocessing预处理" class="headerlink" title="Preprocessing预处理"></a>Preprocessing预处理</h3></li>
</ol>
<ul>
<li>根据方差/均值比（默认值为<strong>0.5</strong>）选择要进行插补的基因</li>
<li>把这些基因分成<strong>N个</strong>随机的亚群，每个亚群有S个基因数（默认情况下，<strong>S设置为512</strong>）</li>
<li>如果目标基因的数目不是这个数目的倍数，我们将要输入的基因的数目<strong>取整</strong>到深度神经网络的N+1子集中<h3 id="Network-architecture构建网络"><a href="#Network-architecture构建网络" class="headerlink" title="Network architecture构建网络"></a>Network architecture构建网络</h3></li>
<li>与目标基因相关的基因<strong>输入层</strong></li>
<li>具有校正线性单元（<strong>ReLU</strong>）激活功能的<strong>256个神经元完全连接的隐藏层</strong></li>
<li>Dropout layer: Dropout是在神经网络中用来解决过拟合问题的一种方法，dropout rate：比率被优化为20%，其他默认参数包括学习率为0.0001，批处理大小为64，子集大小为512。</li>
<li>S目标基因的<strong>输出层</strong><h3 id="Training-procedure"><a href="#Training-procedure" class="headerlink" title="Training procedure"></a>Training procedure</h3>The training starts by splitting the cells between a training (<strong>95%</strong>) and a test set (<strong>5%</strong>).<br>We use a <strong>weighted mean squared error加权均方误差 (MSE) loss function</strong> that gives higher weights to genes with higher expression values.<h3 id="Imputation插补"><a href="#Imputation插补" class="headerlink" title="Imputation插补"></a>Imputation插补</h3></li>
<li>Once the network weights are properly trained, we impute the data by filling zeros in the original matrix with the imputed values.一旦<strong>网络权值</strong>经过适当的训练，我们就可以通过在原始矩阵中填充零来输入数据。</li>
</ul>
<h3 id="数据来源："><a href="#数据来源：" class="headerlink" title="数据来源："></a>数据来源：</h3><p>四个公开可用的<strong>scRNA序列数据集</strong></p>
<h2 id="results"><a href="#results" class="headerlink" title="results"></a>results</h2><h3 id="Mask"><a href="#Mask" class="headerlink" title="Mask"></a>Mask</h3><p>随机掩盖(用零替换) 数据集表达矩阵的一部分，然后测量被掩盖数据的补全值和实际值之间的差异。<br>指标：相关系数和均方差</p>
<h3 id="RNA-FISH（湿实验）"><a href="#RNA-FISH（湿实验）" class="headerlink" title="RNA FISH（湿实验）"></a>RNA FISH（湿实验）</h3><p>对于相同的基因列表，使用其RNA FISH与估算后的基因（或原始scRNA-seq数据）的GINI系数进行比较<br>指标：比较RNA FISH和原始scRNA-seq数据之间的相关系数、均方差<br>结果：DeepImpute最优，DCA其次</p>
<h3 id="聚类"><a href="#聚类" class="headerlink" title="聚类"></a>聚类</h3><p>使用由48,267个来自小鼠的带注释的初级视觉皮层细胞组成实验数据集(Hrvatin)，这些细胞有33个先前的细胞类型标签。使用Scanpy工具包结合Leiden聚类算法进行细胞聚类，<br>评判聚类准确性指标：调整后的兰德指数(ARI)，调整后的相互评(AMS)、福克斯-马洛指数(FMI) 和 轮廓指数(SI)，<br>结论：DeepImpute方法这些指标提升较多，DCA少量提升，其它效果不佳</p>
</div></div><a class="button-hover more" href="2020/04/03/DeepImpute/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/04/01/srf之论文阅读Dr.VAE/">srf之论文阅读Dr.VAE</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-03</time><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="categories/DeepLearning/">DeepLearning</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="tags/srf/">srf</a></div></div><div class="post-content"><div class="main-content content"><h1 id="Dr-VAE-improving-drug-response-prediction-via-modeling-of-drug-perturbation-effects"><a href="#Dr-VAE-improving-drug-response-prediction-via-modeling-of-drug-perturbation-effects" class="headerlink" title="Dr.VAE: improving drug response prediction via modeling of drug perturbation effects"></a>Dr.VAE: improving drug response prediction via modeling of drug perturbation effects</h1><h2 id="1-为什么要讲这个"><a href="#1-为什么要讲这个" class="headerlink" title="1. 为什么要讲这个?"></a>1. 为什么要讲这个?</h2><p><strong>Motivation:</strong><br>将<strong>生物逻辑系统的先验知识</strong>融入到这些方法中是<strong>提高预测性能</strong>的一个很有前途的途径。<br><strong>药物诱导转录组微扰效应的高通量细胞系分析</strong>(High-throughput cell line assays of drug-induced transcriptomic perturbation effects)是一项尚未完全纳入药物反应预测模型的先验知识。</p>
<h2 id="2-提出了什么方法"><a href="#2-提出了什么方法" class="headerlink" title="2. 提出了什么方法 ?"></a>2. 提出了什么方法 ?</h2><p>引入了一个统一的<strong>概率方法</strong>，<strong>Dr.VAE</strong>(Drug Response Variational Autoencoder、药物反应变分自动编码器），它同时根据<strong>活性</strong>和<strong>转录组的扰动</strong>来模拟药物反应。<del>翻译怪怪的??</del> </p>
<p>已知：</p>
<ol>
<li>VAE可以提取出具有生物学意义的癌症经切切片特征</li>
<li>在一种名为DeepProfile的药物反应预测方法中，将预先训练的VAE和单独训练的线性模型相结合。<br>我们旨在共同学习一种潜在的嵌入方法，以提高我们预测药物反应（表型结果）的能力，同时利用最初不受监督的（未知表型结果）药物每次扰动实验来帮助学习这种嵌入方法。</li>
</ol>
<p>放模型图！</p>
<h2 id="3-传统方法有什么问题-（对比）"><a href="#3-传统方法有什么问题-（对比）" class="headerlink" title="3. 传统方法有什么问题?（对比）"></a>3. 传统方法有什么问题?（对比）</h2><p>现有的机器学习算法：<br>深度学习的方法：典型的判别前馈神经网络在药物反应预测中效果不佳，很可能是由于数据限制（特征数量与样本数量）所致。</p>
<p>Result部分<br><em>代码体现</em></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">classif_baseline_perf</span><span class="params">(xtrain, ytrain, xtest, ytest, svmkernel=<span class="string">'rbf'</span>, rseed=None)</span>:</span></span><br><span class="line">    <span class="string">'''这个包括了和 Random Forest、LogisticRegression、SVM、SVM w/ linear kernel的对比'''</span></span><br><span class="line">    data = [xtrain, ytrain, xtest, ytest]</span><br><span class="line">    perf = OrderedDict()</span><br><span class="line"></span><br><span class="line">    <span class="comment">## Random Forest</span></span><br><span class="line">    cl = sklearn.ensemble.RandomForestClassifier(n_estimators=<span class="number">100</span>, random_state=rseed)</span><br><span class="line">    rf_perf = measure_cl_perf(cl, <span class="string">'RF100'</span>, *data)</span><br><span class="line">    perf = concat_dicts(perf, rf_perf)</span><br><span class="line"></span><br><span class="line">	<span class="comment">## LogisticRegression</span></span><br><span class="line">    cl = sklearn.linear_model.LogisticRegressionCV(penalty=<span class="string">'l2'</span>, random_state=rseed)</span><br><span class="line">    lr_perf = measure_cl_perf(cl, <span class="string">'Ridge'</span>, *data)</span><br><span class="line">    perf = concat_dicts(perf, lr_perf)</span><br><span class="line"></span><br><span class="line">    <span class="comment">## SVM </span></span><br><span class="line">    cl = sklearn.svm.SVC(kernel=svmkernel, probability=<span class="literal">True</span>, random_state=rseed)</span><br><span class="line">    svm_perf = measure_cl_perf(cl, <span class="string">'SVM'</span> + svmkernel[:<span class="number">3</span>], *data)</span><br><span class="line">    perf = concat_dicts(perf, svm_perf)</span><br><span class="line">    <span class="comment"># SVM w/ linear kernel</span></span><br><span class="line">    cl = sklearn.svm.SVC(kernel=<span class="string">'linear'</span>, probability=<span class="literal">True</span>, random_state=rseed)</span><br><span class="line">    svm_perf = measure_cl_perf(cl, <span class="string">'SVMlin'</span>, *data)</span><br><span class="line">    perf = concat_dicts(perf, svm_perf)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">return</span> perf</span><br></pre></td></tr></table></figure>

<p>measure_cl_perf()函数作用是评估，计算出的是Acc、AUROC、AUPR<br>concat_dicts()转化为字典</p>
<h2 id="4-原理简单讲一讲（数学方面要求不是很高，有代码实现更方便）"><a href="#4-原理简单讲一讲（数学方面要求不是很高，有代码实现更方便）" class="headerlink" title="4. 原理简单讲一讲（数学方面要求不是很高，有代码实现更方便）?"></a>4. 原理简单讲一讲（数学方面要求不是很高，有代码实现更方便）?</h2><h2 id="5-数据集是怎么划分的"><a href="#5-数据集是怎么划分的" class="headerlink" title="5. 数据集是怎么划分的?"></a>5. 数据集是怎么划分的?</h2><blockquote>
<p><strong>Training data sets</strong>：<strong>CTRPv2</strong>( Cancer Therapeutic Response Portal)  &amp;  <strong>CMap</strong> (CMap-L1000v1)<br><strong>Input data types</strong>：<strong>mRNA</strong> EXP (before &amp; after treatment)</p>
</blockquote>
<h3 id="两个数据集的基本介绍："><a href="#两个数据集的基本介绍：" class="headerlink" title="两个数据集的基本介绍："></a>两个数据集的基本介绍：</h3><p><strong>CTRPv2</strong>：为860个细胞系和481个药物化合物的组合<strong>提供了不同药物浓度下细胞系的相对生存能力</strong>。细胞系对<strong>药物治疗的敏感性</strong>通过<strong>剂量-反应曲线AAC</strong>上方的面积来量化，我们进一步采用瀑布法对连续的AAC进行二值化（Barretina等人，2012；Haibe-Kains等人，2013），<strong>将灵敏度预测任务转化为离散的分类任务。</strong>  <del>体现在哪里？</del><br><strong>CMap</strong>：筛选了19 811种药物化合物对多达77个细胞系中L1000个标志性基因表达的干扰效应。<br>在CMap中的实验并<strong>不能测量药物治疗的敏感性</strong>，但是一些细胞系也在CTRPv2中进行了独立的测试。我们<strong>交叉</strong>引用这些细胞系(取交集)，并将相应的标签分配给它们的扰动测量。</p>
<h3 id="实验数据选择标准："><a href="#实验数据选择标准：" class="headerlink" title="实验数据选择标准："></a>实验数据选择标准：</h3><ol>
<li>从CMap数据集，选择了在<strong>最常见浓度水平下对每种药物进行6小时的微扰实验</strong>。如果细胞系没有在选定的浓度下进行检测，我们使用最接近的检测浓度。接下来，我们将对照组（DMSO载体）实验与药物扰动实验通过批次ID和微珠ID进行匹配，以最小化病例和对照组之间的批次效应。<del>？？</del> 此外，我们通过相关性（&gt;0.75 Pearson q）过滤所选的病例对照对，以过滤出可能标记错误的实验或异常值。</li>
<li>CTRPv2和CMap-L1000v1数据集共有<strong>973</strong>个共同基因。将<strong>每个基因的表达值标准化为零均值和单位方差</strong>。为了进一步的均匀化，包括去除批效应和两个合并数据之间的差异，我们还从集合数据集中去除了第一个主成分（解释了12.8%的变化）。<br>叨叨：数据的筛选真的太讲究了<br>基于以上两个标准，我们选择了<strong>26</strong>种在CTRPv2和CMap-L1000v1数据集中测试的<strong>药物</strong>：</li>
</ol>
<ul>
<li>对于每种选择的药物，至少有<strong>8个不同的细胞系在CMap干扰实验中测试。</strong></li>
<li>CTRPv2中至少有20%的筛选细胞系在剂量-反应AAC二值化后<strong>对药物敏感</strong>。</li>
</ul>
<p><del>插图</del> </p>
<h2 id="7-交叉验证怎么做"><a href="#7-交叉验证怎么做" class="headerlink" title="7. 交叉验证怎么做?"></a>7. 交叉验证怎么做?</h2><p>通过执行20次重复的<strong>5倍交叉验证</strong>，我们生成了100个训练验证测试数据拆分。<br><em>代码体现</em></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">getSplitsByGroupKFold</span><span class="params">(groups, fold, n_splits, shuffle, random_state)</span>:</span></span><br><span class="line">    <span class="string">'''the same group will not appear in two different folds'''</span></span><br><span class="line">    <span class="keyword">assert</span> (n_splits &gt;= <span class="number">3</span>)</span><br><span class="line">    kf = GroupKFold(n_splits=n_splits)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> shuffle:</span><br><span class="line">        <span class="comment"># randomly rename groups so that the GroupKFold (which sorts by group ids first) splits can be randomized</span></span><br><span class="line">        unique_groups = np.unique(groups)</span><br><span class="line">        rnd_renames = sklearn.utils.shuffle(np.arange(len(unique_groups)), random_state=random_state)</span><br><span class="line">        groups_renamed = np.array([rnd_renames[np.argwhere(unique_groups == g)[<span class="number">0</span>]] <span class="keyword">for</span> g <span class="keyword">in</span> groups])</span><br><span class="line">        kfsplit = kf.split(X=np.zeros(groups.shape[<span class="number">0</span>]), groups=groups_renamed)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        kfsplit = kf.split(X=np.zeros(groups.shape[<span class="number">0</span>]), groups=groups)</span><br><span class="line">    allsplits = np.array([x <span class="keyword">for</span> x <span class="keyword">in</span> kfsplit])</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> isinstance(fold, list) <span class="keyword">or</span> isinstance(fold, tuple):</span><br><span class="line">        <span class="keyword">assert</span> (len(fold) == <span class="number">2</span>)</span><br><span class="line">        te_fold = fold[<span class="number">0</span>]</span><br><span class="line">        va_fold = fold[<span class="number">1</span>]</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        te_fold = fold</span><br><span class="line">        va_fold = <span class="number">1</span></span><br><span class="line">    <span class="keyword">assert</span> (te_fold &gt;= <span class="number">1</span> <span class="keyword">and</span> te_fold &lt;= n_splits)      <span class="comment">## the test fold is between 1..n_splits</span></span><br><span class="line">    <span class="keyword">assert</span> (va_fold &gt;= <span class="number">1</span> <span class="keyword">and</span> va_fold &lt;= n_splits - <span class="number">1</span>)  <span class="comment">## the validation fold is between 1..(n_splits-1)</span></span><br><span class="line">    <span class="comment"># the te_fold is indexed from 1 so we need to subtract 1</span></span><br><span class="line">    te_fold_id = te_fold - <span class="number">1</span></span><br><span class="line">    va_fold_id = (te_fold_id + va_fold) % n_splits</span><br><span class="line">    ind_te = allsplits[te_fold_id][<span class="number">1</span>]</span><br><span class="line">    ind_va = allsplits[va_fold_id][<span class="number">1</span>]</span><br><span class="line">    ind_tr = np.concatenate(allsplits[np.setdiff1d(np.arange(n_splits), np.asarray([te_fold_id, va_fold_id]))][:, <span class="number">1</span>])</span><br><span class="line">    <span class="keyword">assert</span> (checkSplit(ind_tr, ind_va, ind_te))</span><br><span class="line">    <span class="keyword">return</span> ind_tr, ind_va, ind_te</span><br></pre></td></tr></table></figure>

<h2 id="8-Discussion："><a href="#8-Discussion：" class="headerlink" title="8. Discussion："></a>8. Discussion：</h2><p>典型的判别前馈神经网络在药物反应预测中效果不佳，很可能是由于数据限制（特征数量与样本数量）所致。 我们表明，<strong>药物反应和扰动效应的联合生成模型</strong>在很大程度上缓解了这种情况，<strong>可能起到了有效的正则化和鲁棒的特征提取的作用</strong>，而不会产生过拟合的情况。</p>
<h3 id="局限性："><a href="#局限性：" class="headerlink" title="局限性："></a>局限性：</h3><ol>
<li>只考虑基因表达方式，<strong>没有额外整合</strong>甲基化、拷贝数变化、相互作用状态的<strong>多组预测因子。</strong></li>
<li>我们在每种药物最常见的浓度水平下，对治疗6小时后的CMap-L1000v1扰动进行建模。有人认为，<strong>6小时太短，许多反馈调节机制无法体现自己，因此这些实验本身并不能提供完整的转录反应图片</strong>。值得注意的是，药物细胞系活力测定通常在较长的治疗时间（如72小时）内完成。因此，我们也用24小时的微扰实验来训练我们的VAE博士，然而，<strong>由于这样的实验数量有限</strong>，这并没有提高我们的预测性能。</li>
<li>Dr.VAE所组成的每一个条件分布都是由一个神经网络参数化的。调整超参数以匹配数据复杂性的能力使Dr.VAE成为一个非常灵活的模型。<strong>由于我们选择了简单性，我们的神经网络大多有一个隐藏层，而分类后验函数和扰动函数是线性的。</strong> 然而，到目前为止，我们试图通过更复杂的分布使用更深的网络或使用规范化流来近似后验的尝试，<strong>并没有显著地提高性能来证明增加的复杂性。</strong></li>
</ol>
<hr>
<h2 id="预备-amp-补充知识："><a href="#预备-amp-补充知识：" class="headerlink" title="预备&amp;补充知识："></a>预备&amp;补充知识：</h2><h3 id="什么是VAE-Variational-auto-encoder、变分自编码器-？"><a href="#什么是VAE-Variational-auto-encoder、变分自编码器-？" class="headerlink" title="什么是VAE(Variational auto-encoder、变分自编码器)？"></a>什么是VAE(Variational auto-encoder、变分自编码器)？</h3><p>一个可以和GAN相媲美的生成网络。<br>VAE通过说我希望生成一张新的猫脸，那么这个n维向量代表的就是n个决定最终猫脸模样的隐形因素。对于每个因素，都对应产生一种分布，从这些分布关系中进行采样，那么我就可以通过一个深度网络恢复出最终的猫脸。<br>通过一个网络，将一个高维空间的人脸映射为低维空间的一个向量。那么如果我将这个向量定义为四维，采用one-hot的表达方式表达四张不同的脸，那么这个网络就可以表达四个脸，你输入不同的数据，他就会输出不同的脸来。</p>
<h4 id="VAE有什么用？"><a href="#VAE有什么用？" class="headerlink" title="VAE有什么用？"></a>VAE有什么用？</h4><p>VAE到底有什么用呢，就为了把大量的数据存储到网络结构中去呀，当然不能是这么简单的一个事。想象一下，在一个末日大片中，我们需要去搞出一个许多机器人占领世界的场面，但是我们的特效师一张张去画，这无疑会是一个令人头疼的工作，利用VAE，只需要使用有限的数据输入，我们就可以通过隐形参数的调节，获得几乎无限的数量。</p>
<h4 id="VAE网络结构组成："><a href="#VAE网络结构组成：" class="headerlink" title="VAE网络结构组成："></a>VAE网络结构组成：</h4><p>可以大致分成<strong>Encoder</strong>和<strong>Decoder</strong>两部分。对于输入图片，Encoder将提取得到编码：一个<strong>mean vector</strong>（均值）和一个<strong>deviation vector</strong>（方差），然后将这个编码（两个vector）作为Decoder的输入，最终输出一张和原图相近的图片。</p>
<p>它本质上就是在我们常规的自编码器的基础上，对 encoder 的结果（在VAE中对应着计算均值的网络）加上了<strong>“高斯噪声”</strong>，使得结果 decoder 能够对噪声有鲁棒性；而那个额外的 KL loss（目的是让均值为 0，方差为 1），事实上就是相当于对 encoder 的一个正则项，希望 encoder 出来的东西均有零均值。</p>
<p>drug perturbation effects是什么?</p>
<h3 id="ablation-study消融实验"><a href="#ablation-study消融实验" class="headerlink" title="ablation study消融实验:"></a>ablation study消融实验:</h3><p>论证是什么起了作用，类似于中学学的控制变量<br>模型简化测试。<br>看看取消掉一些模块后性能有没有影响。</p>
<h3 id="参考："><a href="#参考：" class="headerlink" title="参考："></a>参考：</h3><h4 id="文章："><a href="#文章：" class="headerlink" title="文章："></a>文章：</h4><p><a href="https://blog.csdn.net/weixin_40955254/article/details/82315224" target="_blank" rel="noopener">https://blog.csdn.net/weixin_40955254/article/details/82315224</a><br><a href="https://blog.csdn.net/cjh_jinduoxia/article/details/84995156" target="_blank" rel="noopener">https://blog.csdn.net/cjh_jinduoxia/article/details/84995156</a></p>
<h4 id="视频："><a href="#视频：" class="headerlink" title="视频："></a>视频：</h4><p>（白板推导）<a href="https://www.bilibili.com/video/BV15E411w7Pz" target="_blank" rel="noopener">https://www.bilibili.com/video/BV15E411w7Pz</a></p>
</div></div><a class="button-hover more" href="2020/04/01/srf之论文阅读Dr.VAE/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/03/29/srf之论文汇报流程/">srf之论文汇报流程</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-03</time><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="categories/DeepLearning/">DeepLearning</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="tags/srf/">srf</a></div></div><div class="post-content"><div class="main-content content"><h2 id="论文将要讲述的方面"><a href="#论文将要讲述的方面" class="headerlink" title="论文将要讲述的方面"></a>论文将要讲述的方面</h2><ol>
<li>为什么要讲这个?（背景）</li>
<li>提出了什么 <strong>方法</strong> ?</li>
<li>传统方法有什么问题?（<strong>对比</strong>）</li>
<li><strong>原理</strong>简单讲一讲（数学方面要求不是很高，有代码实现更方便）?</li>
<li><strong>数据集</strong>是怎么<strong>划分</strong>的?</li>
<li><strong>交叉验证</strong>怎么做?</li>
</ol>
<h2 id="研究方法和主题"><a href="#研究方法和主题" class="headerlink" title="研究方法和主题"></a>研究方法和主题</h2><p>学长说在GCN方面不会有更深入的探寻了，会在以下几个方面找突破口</p>
<ul>
<li>Attention机制</li>
<li>知识图谱</li>
<li>异构图网络的学习</li>
<li>嵌入式学习</li>
</ul>
<h2 id="论文的选取"><a href="#论文的选取" class="headerlink" title="论文的选取"></a>论文的选取</h2><blockquote>
<p><strong>新</strong>（前沿），在期刊上发表的更好<br>有代码最好，有名字的模型在github上可以搜索的到</p>
</blockquote>
<h2 id="学长目前的想法"><a href="#学长目前的想法" class="headerlink" title="学长目前的想法"></a>学长目前的想法</h2><ul>
<li>SMILES收集之后：通过label encoding提取smiles的embedding<br><img src="srf%E4%B9%8B%E8%AE%BA%E6%96%87%E6%B1%87%E6%8A%A5%E6%B5%81%E7%A8%8B/%E6%B5%81%E7%A8%8B%E5%9B%BE.png" alt></li>
</ul>
</div></div><a class="button-hover more" href="2020/03/29/srf之论文汇报流程/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/02/26/web小项目/">Web小项目(react调试和部署)</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-03-29</time></div><div class="post-content"><div class="main-content content"><h2 id="如何在本地调试react项目"><a href="#如何在本地调试react项目" class="headerlink" title="如何在本地调试react项目"></a>如何在本地调试react项目</h2><blockquote>
<p>在项目终端下(pycharm)运行 <figure class="highlight plain"><figcaption><span>install```**安装依赖**</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">项目会生成**node_modules**目录</span><br><span class="line"></span><br><span class="line">我们在在项目终端下(pycharm)运行</span><br></pre></td></tr></table></figure></p>
</blockquote>
<p>npm start</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">浏览器会自动打开http://localhost:3000运行调试</span><br><span class="line"></span><br><span class="line">## 如何将本地的react项目部署到云服务器上</span><br><span class="line">我们在在项目终端下(pycharm)运行</span><br></pre></td></tr></table></figure>

<p>npm run build</p>
<pre><code>项目会生成了一个**build**目录

打开宝塔服务器页面，到相关网站下，将build页面下==static上传，以及和static平行的文件上传。==

至此，大功告成！
tips:感觉网上的教程大多数很复杂，暂时也不太明白他们做的原理(TODO)</code></pre></div></div><a class="button-hover more" href="2020/02/26/web小项目/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/02/25/爬虫之selenium报错——Chromedriver/">爬虫之selenium报错——Chromedriver</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-02</time></div><div class="post-content"><div class="main-content content"><p><strong>报错信息：</strong></p>
<figure class="highlight plain"><figcaption><span>executable needs to be in PATH.```</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">**解决方法：**</span><br><span class="line">### 关于Chromedriver如何配置环境变量问题解决[^1]</span><br><span class="line">到环境变量中加上‪C:\Program Files (x86)\Google\Chrome\Application</span><br><span class="line"></span><br><span class="line">### windows下更新环境变量不需要重启系统快速生效的方法[^2]</span><br></pre></td></tr></table></figure>

<p>set path=test<br>echo %path%<br>```</p>
<h3 id="如果还没有生效的话-3"><a href="#如果还没有生效的话-3" class="headerlink" title="如果还没有生效的话[^3]"></a>如果还没有生效的话[^3]</h3><p><strong>原来还要放到Python目录下</strong><br>很多人都反应上面两步做了也没有生效</p>
<p>最后……<br>自己入自己的坑，才是最绝望的。chromedriver和chrome 多少是眼残的事……</p>
<p>参考：<br>[^1]: <a href="https://blog.csdn.net/qq_41429288/article/details/80472064" target="_blank" rel="noopener">https://blog.csdn.net/qq_41429288/article/details/80472064</a><br>[^2]: <a href="https://blog.csdn.net/u010770041/article/details/49915089" target="_blank" rel="noopener">https://blog.csdn.net/u010770041/article/details/49915089</a><br>[^3]: <a href="https://blog.csdn.net/weixin_37185329/article/details/80493281" target="_blank" rel="noopener">https://blog.csdn.net/weixin_37185329/article/details/80493281</a></p>
</div></div><a class="button-hover more" href="2020/02/25/爬虫之selenium报错——Chromedriver/#more">阅读全文</a></div><div class="recent-post-item"><a class="post-title" href="2020/02/23/ideaVim-easymotion 插件/">ideaVim-easymotion 插件</a><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 更新于 2020-04-02</time></div><div class="post-content"><div class="main-content content"><h4 id="作用：辅助（优化）ideaVim的"><a href="#作用：辅助（优化）ideaVim的" class="headerlink" title="作用：辅助（优化）ideaVim的"></a>作用：辅助（优化）ideaVim的</h4><p>主要是通过提升了<strong>跳转</strong>方式<br>一般的：<br><code>gg, G</code>最上、最下<br><code>Ctrl-D/U</code>下几步，上几步</p>
<p>网上教程不多，具体更多详细教程：<br><a href="http://www.wklken.me/posts/2015/06/07/vim-plugin-easymotion.html" target="_blank" rel="noopener">http://www.wklken.me/posts/2015/06/07/vim-plugin-easymotion.html</a></p>
</div></div><a class="button-hover more" href="2020/02/23/ideaVim-easymotion 插件/#more">阅读全文</a></div></div><div id="pagination"><div class="pagination"><span class="page-number current">1</span><a class="page-number" href="page/2/">2</a><span class="space">&hellip;</span><a class="page-number" href="page/4/">4</a><a class="extend next" rel="next" href="page/2/"><i class="fas fa-angle-right"></i></a></div></div></div></div><div class="button-hover" id="return-top"><i class="fas fa-arrow-up" aria-hidden="true"></i></div><footer><div id="footer"><div class="button-hover" id="side-button"><i class="fas fa-arrow-right"></i></div><div class="right-content"><div class="busuanzi"><script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><span id="busuanzi_container_site_uv"><i class="fas fa-user"></i></span><span id="busuanzi_value_site_uv"></span><span></span><span class="footer-separator">|</span><span id="busuanzi_container_site_pv"><i class="fas fa-eye"></i></span><span id="busuanzi_value_site_pv"></span><span></span></div><div class="copyright">&copy;2017 ～ 2020 By Rr-shan</div></div></div></footer></div><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/thirdparty/jquery-3.3.1.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/thirdparty/velocity.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/thirdparty/jquery.mCustomScrollbar.concat.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/fan.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/canvas_bg.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/utils.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/scroll.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/sidebar.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="/js/copy.js"></script><!--script(src=url)--><script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"debug":false,"model":{"jsonPath":"/live2dw/assets/assets/haruto.model.json"},"display":{"position":"left","width":150,"height":300},"mobile":{"show":true},"log":false});</script></body></html>