<!DOCTYPE html><html lang="zh-CN"><head><meta name="generator" content="Hexo 3.9.0"><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"><meta name="description" content="论文阅读DrImpute"><meta name="keywords" content="Dropouts课题(刘世超老师)"><meta name="author" content="Rr-shan,undefined"><meta name="copyright" content="Rr-shan"><title>论文阅读DrImpute【Sanzzi】</title><link rel="stylesheet" href="../../../../css/fan.css"><link rel="stylesheet" href="../../../../css/thirdparty/jquery.mCustomScrollbar.min.css"><link rel="icon" href="../../../../favicon.ico"><!-- script(src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML")--><script src="../../../../js/mathjax/mathjax.js"></script><script type="text/x-mathjax-config">MathJax.Hub.Config({
    tex2jax: {inlineMath: [['$', '$'], ['\\(', '\\)']]}
});
</script><script>var isPassword = '' || false;
if (isPassword) {
    if (prompt('请输入文章密码') !== '') {
        alert('密码错误！');
        history.back();
    }
}</script><script>window.GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  }
}</script></head><body><canvas id="universe"></canvas><!--#body--><div id="sidebar"><div class="toggle-sidebar-info button-hover"><span data-toggle="文章目录">站点概览</span></div><div class="sidebar-toc"><div class="sidebar-toc-title">目录</div><div class="sidebar-toc-progress"><span class="progress-notice">您已阅读了</span><span class="progress-num">0</span><span class="progress-percentage">%</span><div class="sidebar-toc-progress-bar"></div></div><div class="sidebar-toc-content" id="sidebar-toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#原理"><span class="toc-number">1.</span> <span class="toc-text">原理</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#聚类"><span class="toc-number">1.0.1.</span> <span class="toc-text">聚类</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#1-preprocess：data-cleansing-normalization-and-log-transformation"><span class="toc-number">1.0.2.</span> <span class="toc-text">1. preprocess：data cleansing, normalization, and log transformation</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#2-calculating-the-distance-matrix-among-cells"><span class="toc-number">1.0.3.</span> <span class="toc-text">2.  calculating the distance matrix among cells</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#3-imputing-the-dropout-entries-based-on-the-clustering-results"><span class="toc-number">1.0.4.</span> <span class="toc-text">3.  imputing the dropout entries based on the clustering results</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#4-averaging-all-imputation-results-to-determine-the-final-imputation"><span class="toc-number">1.0.5.</span> <span class="toc-text">4.  averaging all imputation results to determine the final imputation.</span></a></li></ol></li></ol><li class="toc-item toc-level-2"><a class="toc-link" href="#实验"><span class="toc-number">2.</span> <span class="toc-text">实验</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Experiment-1："><span class="toc-number">2.1.</span> <span class="toc-text">Experiment 1：</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Experiment-2"><span class="toc-number">2.2.</span> <span class="toc-text">Experiment 2</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Experiment-3"><span class="toc-number">2.3.</span> <span class="toc-text">Experiment 3</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Experiment-4"><span class="toc-number">2.4.</span> <span class="toc-text">Experiment 4</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#数据集："><span class="toc-number">3.</span> <span class="toc-text">数据集：</span></a></li></div></div><div class="author-info hide"><div class="author-info-avatar"><img class="author-info-avatar-img" src="/avatar.png"></div><div class="author-info-name">Rr-shan</div><div class="author-info-description"></div><div class="links-buttons"><a class="links-button button-hover" href="https://github.com/Rr-shan" target="_blank">GitHub<i class="icon-dot bg-color3"></i></a><a class="links-button button-hover" href="mailto:1224559633@qq.com" target="_blank">E-Mail<i class="icon-dot bg-color7"></i></a></div><div class="author-info-articles"><a class="author-info-articles-archives article-meta" href="../../../../archives"><span class="pull-top">日志</span><span class="pull-bottom">46</span></a><a class="author-info-articles-tags article-meta" href="../../../../tags"><span class="pull-top">标签</span><span class="pull-bottom">12</span></a><a class="author-info-articles-categories article-meta" href="../../../../categories"><span class="pull-top">分类</span><span class="pull-bottom">12</span></a></div><div class="friend-link"><a class="friend-link-text" href="https://chao-yin-github.github.io/" target="_blank">yccccc~~~~~~~~</a><a class="friend-link-text" href="https://sanzzi.cn" target="_blank">Mywebsite</a></div></div></div><div id="main-container"><header><div id="menu-outer"><i class="menu-list-icon fas fa-bars"></i><nav id="menu-inner"><a class="menu-item" href="/">首页</a><a class="menu-item" href="/tags">标签</a><a class="menu-item" href="/categories">分类</a><a class="menu-item" href="/archives">归档</a><a class="menu-item" href="/about">关于</a></nav><div class="right-info"><a class="title-name" href="/">Sanzzi</a><span id="now-time"></span></div></div></header><div id="content-outer"><div id="content-inner"><article id="post"><div class="post-header"><div class="title">论文阅读DrImpute</div><div class="container"><time class="button-hover post-date"><i class="fas fa-calendar-alt article-icon" aria-hidden="true"></i> 发表于 2019-12-11 | 更新于 2020-04-05</time><!--time.button-hover.post-date #[i.fas.fa-calendar-alt.article-icon(aria-hidden="true")] #[=__('post.modified')] #[=date(page['updated'], config.date_format)]--><div class="button-hover categories"><i class="fa fa-inbox article-icon" aria-hidden="true"></i><a class="link-a" href="../../../../categories/ML-DL/">ML&amp;DL</a></div><div class="button-hover tags"><i class="fa fa-tag article-icon" aria-hidden="true"></i><a class="link-a" href="../../../../tags/Dropouts课题-刘世超老师/">Dropouts课题(刘世超老师)</a></div></div></div><div class="main-content"><h2 id="原理"><a href="#原理" class="headerlink" title="原理"></a>原理</h2><h4 id="聚类"><a href="#聚类" class="headerlink" title="聚类"></a>聚类</h4><h4 id="1-preprocess：data-cleansing-normalization-and-log-transformation"><a href="#1-preprocess：data-cleansing-normalization-and-log-transformation" class="headerlink" title="1. preprocess：data cleansing, normalization, and log transformation"></a>1. <strong>preprocess：</strong>data cleansing, normalization, and log transformation</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">preprocess &lt;- <span class="keyword">function</span>(x, min.expressed.gene = <span class="number">0</span>, min.expressed.cell = <span class="number">2</span>, max.expressed.ratio = <span class="number">1</span>, normalize.by.size.effect = <span class="literal">FALSE</span>)&#123;</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> (class(x) == <span class="string">'SummarizedExperiment'</span>)</span><br><span class="line">		X &lt;- assays(x)$count</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (class(x) == <span class="string">'matrix'</span>)</span><br><span class="line">		X &lt;- x</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (is(x, <span class="string">'sparseMatrix'</span>))</span><br><span class="line">		X &lt;- x</span><br><span class="line">	<span class="keyword">else</span></span><br><span class="line">		<span class="keyword">stop</span>(sprintf(<span class="string">'unknown class(x): %s'</span>, class(x)))</span><br><span class="line"></span><br><span class="line">	M &lt;- ncol(X)</span><br><span class="line">	N &lt;- nrow(X)</span><br><span class="line">	m &lt;- Matrix::colSums(X &gt; <span class="number">1</span>) &gt;= min.expressed.gene	<span class="comment"># cells that have at least min.expressed.gene expreseed genes</span></span><br><span class="line">	n &lt;- Matrix::rowSums(X &gt; <span class="number">1</span>) &lt;= max.expressed.ratio * M &amp; Matrix::rowSums(X &gt; <span class="number">1</span>) &gt;= min.expressed.cell	<span class="comment"># genes that are detected in at least min.expressed.cell or at most max.expressed.ratio cells</span></span><br><span class="line">	<span class="keyword">if</span> (normalize.by.size.effect)&#123;</span><br><span class="line">	  sf &lt;- apply((X[n, m] + <span class="number">1</span>) / exp(Matrix::rowMeans(log(X[n, m] + <span class="number">1</span>))), <span class="number">2</span>, median)</span><br><span class="line">		X &lt;- t(t(X[n, m]) / sf)</span><br><span class="line">	&#125;<span class="keyword">else</span></span><br><span class="line">		X &lt;- X[n, m]</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> (class(x) == <span class="string">'SummarizedExperiment'</span>)&#123;</span><br><span class="line">		x &lt;- x[n, m]</span><br><span class="line">		assays(x)$count &lt;- X</span><br><span class="line">	&#125;<span class="keyword">else</span> <span class="keyword">if</span> (class(x) == <span class="string">'matrix'</span>)&#123;</span><br><span class="line">		x &lt;- as.matrix(X)</span><br><span class="line">	&#125;<span class="keyword">else</span> <span class="keyword">if</span> (is(x, <span class="string">'sparseMatrix'</span>))&#123;</span><br><span class="line">		x &lt;- X</span><br><span class="line">	&#125;</span><br><span class="line">	</span><br><span class="line">	x</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="2-calculating-the-distance-matrix-among-cells"><a href="#2-calculating-the-distance-matrix-among-cells" class="headerlink" title="2.  calculating the distance matrix among cells"></a>2.  calculating the distance matrix among cells</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">dist2 &lt;- <span class="keyword">function</span>(X, method)&#123;</span><br><span class="line">	<span class="keyword">if</span> (method == <span class="string">'spearman'</span>)</span><br><span class="line">		D &lt;- as.matrix(<span class="number">1</span> - cor(as.matrix(X), method = <span class="string">'spearman'</span>))</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (method == <span class="string">"pearson"</span>)</span><br><span class="line">		D &lt;- as.matrix(<span class="number">1</span> - cor(as.matrix(X), method = <span class="string">'pearson'</span>))</span><br><span class="line">	<span class="keyword">else</span> <span class="keyword">if</span> (method == <span class="string">"euclidean"</span>) </span><br><span class="line">		D &lt;- as.matrix(dist(t(as.matrix(X)), method = <span class="string">'euclidean'</span>))</span><br><span class="line">	<span class="keyword">else</span></span><br><span class="line">		<span class="keyword">stop</span>(sprintf(<span class="string">'unknown method: %s'</span>, method))</span><br><span class="line">	D</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="3-imputing-the-dropout-entries-based-on-the-clustering-results"><a href="#3-imputing-the-dropout-entries-based-on-the-clustering-results" class="headerlink" title="3.  imputing the dropout entries based on the clustering results"></a>3.  imputing the dropout entries based on the clustering results</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">DrImpute &lt;- <span class="keyword">function</span>(X, ks = <span class="number">10</span>:<span class="number">15</span>, dists = c(<span class="string">'spearman'</span>, <span class="string">'pearson'</span>), batch.size = <span class="literal">NA</span>, mc.cores = <span class="number">1</span>)&#123;</span><br><span class="line"></span><br><span class="line">	N &lt;- nrow(X)	<span class="comment"># number of genes</span></span><br><span class="line">	M &lt;- ncol(X) 	<span class="comment"># number of cells</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] number of cells: %d\n'</span>, Sys.time(), M))</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] number of genes: %d\n'</span>, Sys.time(), N))</span><br><span class="line"></span><br><span class="line">	is.zero &lt;- as(X == <span class="number">0</span>, <span class="string">'lgeMatrix'</span>)	<span class="comment"># dense logical matrix for zero and weakly detected events # 零和弱检测事件的密集逻辑矩阵</span></span><br><span class="line"></span><br><span class="line">	is.expressed &lt;- as(X &gt; log(<span class="number">1</span> + <span class="number">5</span>), <span class="string">'lgeMatrix'</span>) <span class="comment"># dense logical matrix for robustly expressed events # 鲁棒表示事件的密集逻辑矩阵</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% zero events (X=0): %.1f%%\n'</span>, Sys.time(), sum(is.zero) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% robustly expressed events (X&gt;5): %.1f%%\n'</span>, Sys.time(), sum(is.expressed) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	<span class="keyword">if</span> (!is.na(batch.size))&#123;</span><br><span class="line">		cat(sprintf(<span class="string">'[%s] batch size: %d cells\n'</span>, Sys.time(), batch.size))</span><br><span class="line">	&#125;</span><br><span class="line">	cat(sprintf(<span class="string">'[%s] # of CPU cores: %d\n'</span>, Sys.time(), mc.cores))</span><br><span class="line"></span><br><span class="line">	X &lt;- estimate.expression(X, ks = ks, dists = dists, mc.cores = mc.cores, batch.size = batch.size)</span><br><span class="line">	is.zero &lt;- as(X == <span class="number">0</span>, <span class="string">'lgeMatrix'</span>)	<span class="comment"># dense logical matrix for zero indicators</span></span><br><span class="line">	cat(sprintf(<span class="string">'[%s] %% zero events after imputation: %.1f%%\n'</span>, Sys.time(), sum(is.zero) / (N * M) * <span class="number">100</span>))</span><br><span class="line">	X</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h4 id="4-averaging-all-imputation-results-to-determine-the-final-imputation"><a href="#4-averaging-all-imputation-results-to-determine-the-final-imputation" class="headerlink" title="4.  averaging all imputation results to determine the final imputation."></a>4.  averaging all imputation results to determine the final imputation.</h4><h2 id="实验"><a href="#实验" class="headerlink" title="实验"></a>实验</h2><h3 id="Experiment-1："><a href="#Experiment-1：" class="headerlink" title="Experiment 1："></a>Experiment 1：</h3><p><strong>Method:</strong><br>the <strong>cell-cell</strong> distance matrix was computed using <strong>Spearman and Pearson correlations</strong>, followed by the <strong>cell-wise clustering</strong> based upon the distance matrix over a range of expected number of clusters k (k ranging from 10 to 15 by default).</p>
<p><strong>Evaluation indicators:</strong><br>To generate the dropout zero, we randomly <strong>down-sampled</strong> the raw sequencing reads to 10, 15, 25, 40 and 63%f the total number of reads</p>
<p><strong>Evaluation indicators:<br>F1 score</strong></p>
<h3 id="Experiment-2"><a href="#Experiment-2" class="headerlink" title="Experiment 2"></a>Experiment 2</h3><p><strong>Method:</strong><br>The effect of cell clustering.<br> For each dataset, we randomly sampled 100 genes (<strong>down-sampling</strong>), or one-third of the total cells (down-sampling), and we clustered the cells using each of the clustering methods with and without preprocessing the down-sampled dataset using DrImpute. This process was <strong>repeated 100</strong> times, and we compared how consistent the clustering results were after down-sampling the genes or cells as measured by cross ARI<br><strong>Evaluation indicators:<br>ARI</strong></p>
<h3 id="Experiment-3"><a href="#Experiment-3" class="headerlink" title="Experiment 3"></a>Experiment 3</h3><p><strong>Method:</strong><br>To evaluate the accuracy of the <strong>dimension reduction</strong> in 2D space, we first estimated how discriminatively the cells from one population (using the class label reported in the original publication) separated from other populations in 2D space. For each dimension reduction result</p>
<p><strong>Evaluation indicators:</strong><br>we used the 2D coordinates of 90% of cells as the feature to train a linear support vector machine (SVM) classifier,and we predicted the class label for the remaining 10%of the cells. The above process was repeated ten times,and the overall prediction accuracy (10-fold cross validation accuracy) was used to quantitatively measure the separation of different populations in 2D space.// 这块没有讲清楚</p>
<h3 id="Experiment-4"><a href="#Experiment-4" class="headerlink" title="Experiment 4"></a>Experiment 4</h3><p>重建谱系轨迹，推断单细胞的分化和祖细胞状态。(老师也不是很明白)</p>
<h2 id="数据集："><a href="#数据集：" class="headerlink" title="数据集："></a>数据集：</h2><p>scRNA-seq datasets</p>
<p><strong>cell labels</strong><br>“gold”：experimental conditions or cell lines（细胞系）<br>“silver”： derived and assigned by the authors’ knowledge of the underlying biology.<br>“bronze”:  developmental stages (time labels)</p>
</div><div class="post-copyright"><div class="post-copyright-author"><span class="post-copyright-meta">本文作者: </span><span class="post-copyright-info"><a href="mailto:undefined">Rr-shan</a></span></div><div class="post-copyright-type"><span class="post-copyright-meta">本文链接: </span><span class="post-copyright-info"><a href="https://rr-shan.github.io/2019/12/11/论文阅读DrImpute/">https://rr-shan.github.io/2019/12/11/论文阅读DrImpute/</a></span></div><div class="post-copyright-notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://rr-shan.github.io">Sanzzi</a>！</span></div></div></article><div id="pagination"><div class="prev-post pull-left"><span class="line line-top"></span><span class="line line-right"></span><span class="line line-bottom"></span><span class="line line-left"></span><a href="../../18/DeepImpute/"><i class="fas fa-angle-left">&nbsp;</i><span>论文阅读DeepImpute</span></a></div><div class="next-post pull-right"><span class="line line-top"></span><span class="line line-right"></span><span class="line line-bottom"></span><span class="line line-left"></span><a href="../../09/JShell容易出现错误之处/"><span>JShell容易出现错误之处</span><span>&nbsp;</span><i class="fas fa-angle-right"></i></a></div></div><!--div!= paginator()--></div></div><div class="button-hover" id="return-top"><i class="fas fa-arrow-up" aria-hidden="true"></i></div><footer><div id="footer"><div class="button-hover" id="side-button"><i class="fas fa-arrow-right"></i></div><div class="right-content"><div class="busuanzi"><script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><span id="busuanzi_container_page_pv"><i class="fas fa-file-o"></i></span><span id="busuanzi_value_page_pv"></span><span></span></div><div class="copyright">&copy;2017 ～ 2020 By Rr-shan</div></div></div></footer></div><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/thirdparty/jquery-3.3.1.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/thirdparty/velocity.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/thirdparty/jquery.mCustomScrollbar.concat.min.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/fan.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/canvas_bg.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/utils.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/scroll.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/sidebar.js"></script><!--script(src=url)--><!--js(src=url_for(url) + '?version=' + version())--><script src="../../../../js/copy.js"></script><!--script(src=url)--><script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"debug":false,"model":{"jsonPath":"/live2dw/assets/assets/haruto.model.json"},"display":{"position":"left","width":150,"height":300},"mobile":{"show":true},"log":false});</script></body></html>